lexxpluss_apps/src/rosserial_actuator.hpp

```diff
@@ -26,66 +26,84 @@
-#include "ros/node_handle.h"
-#include "std_msgs/Float32MultiArray.h"
-#include "std_msgs/Int32MultiArray.h"
-#include "lexxauto_msgs/LinearActuatorControlArray.h"
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/can.h>
 #include "actuator_controller.hpp"
 
+#define CAN_ID_ACTUATOR_CONTROL 0x208 //based on saito-san's CAN ID assignment
+#define CAN_ID_ACTUATOR_ENCODER 0x209 //based on saito-san's CAN ID assignment
+#define CAN_ID_ACTUATOR_CURRENT 0x20a //based on saito-san's CAN ID assignment
+#define CAN_DATALENGTH_ACTUATOR_ENCODER 6
+#define CAN_DATALENGTH_ACTUATOR_CURRENT 8
+
+
+
 namespace lexxhard {
 
+char __aligned(4) msgq_can_actuator_control_buffer[8 * sizeof(actuator_controller::can_format_control)];
+k_msgq msgq_can_actuator_control;
+
 class ros_actuator {
 public:
-    void init(ros::NodeHandle &nh) {
-        nh.advertise(pub_encoder);
-        nh.advertise(pub_connection);
-        nh.advertise(pub_current);
-        nh.subscribe(sub_control);
-        msg_encoder.data = msg_encoder_data;
-        msg_encoder.data_length = sizeof msg_encoder_data / sizeof msg_encoder_data[0];
-        msg_connection.data = msg_connection_data;
-        msg_connection.data_length = sizeof msg_connection_data / sizeof msg_connection_data[0];
-        msg_current.data = msg_current_data;
-        msg_current.data_length = sizeof msg_current_data / sizeof msg_current_data[0];
+    int init() {
+        //can device bind
+        k_msgq_init(&msgq_can_actuator_control, msgq_can_actuator_control_buffer, sizeof(actuator_controller::can_format_control), 8);
+        dev = device_get_binding("CAN_2");
+        if (!device_is_ready(dev))
+            return -1;
+
+        //setup can filter
+        static const zcan_filter filter_actuator_control{
+            .id{CAN_ID_ACTUATOR_CONTROL},
+            .rtr{CAN_DATAFRAME},
+            .id_type{CAN_STANDARD_IDENTIFIER},
+            .id_mask{0x7ff},
+            .rtr_mask{1}
+        };
+        can_attach_msgq(dev, &msgq_can_actuator_control, &filter_actuator_control);
+        return 0;
     }
     void poll() {
+        //send to IPC of sensor informations
         actuator_controller::msg message;
         while (k_msgq_get(&actuator_controller::msgq, &message, K_NO_WAIT) == 0) {
+
+
             // ROS:[center,left,right], ROBOT:[left,center,right]
-            msg_encoder.data[0] = message.encoder_count[1];
-            msg_encoder.data[1] = message.encoder_count[0];
-            msg_encoder.data[2] = message.encoder_count[2];
-            pub_encoder.publish(&msg_encoder);
-            msg_connection.data[0] = message.connect * 1e-3f;
-            pub_connection.publish(&msg_connection);
-            msg_current.data[0] = message.current[1] * 1e-3f;
-            msg_current.data[1] = message.current[0] * 1e-3f;
-            msg_current.data[2] = message.current[2] * 1e-3f;
-            pub_current.publish(&msg_current);
+            zcan_frame can_frame_actuator_encoder{
```

Comment:
Since we are switching to V7, this is an opportunity to consider aligning ROS and ROBOT.
It will be difficult to reorder V6, but V7 may be doable.

( I would like to maintain compatibility with V6, so changing ROS is difficult. could we consider if we could change the ROBOT side? )

V7に切り替わるので、ROSとROBOTの並びを揃えるか検討できるチャンスだと思います。
V6の並び替えは難しいでしょうが、V7であればやりようはあるかもしれません。

( V6との互換性を維持したいので、ROSを変えるのは難しい。ROBOT側を変えられないか検討できないでしょうか? )